### PR TITLE
Added implementation

### PR DESCRIPTION
**Pull request by iceseer**
Signed-off-by: iceseer <iceseer@gmail.com>

C bindings over RUST implementations of [reed-solomon-novelpoly](https://github.com/paritytech/reed-solomon-novelpoly) that implements [Novel Polynomial Basis and Its Application to
Reed-Solomon Erasure Codes](https://www.citi.sinica.edu.tw/papers/whc/4454-F.pdf).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+cmake-build-*/
+.idea/
+
+# C headers generated by cbindgen
+include/
+
+# Remove Cargo.lock from gitignore if creating an executable, leave it for libraries
+# More information here https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html
+Cargo.lock
```
Comment (Harrm) [inline .gitignore]: Hm, seeing Cargo.lock both in the commit and in the .gitignore is confusing.

File: src/erasure_coding.rs [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+use std::slice;
+
+const MAX_VALIDATORS: usize = novelpoly::f2e16::FIELD_SIZE;
+
+/// Errors in erasure coding.
+#[repr(C)]
+#[derive(Debug, Clone, PartialEq)]
+pub enum NPRSError {
+    /// No error
+    Ok,
```
Comment (Harrm) [inline src/erasure_coding.rs:12]: NPRSError::Ok looks pretty clumsy. I sugget NPRSResult.

File: src/erasure_coding.rs [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    NonUniformChunks,
+    /// An uneven byte-length of a shard is not valid for `GF(2^16)` encoding.
+    UnevenLength,
+    /// Chunk index out of bounds.
+    ChunkIndexOutOfBounds {
+        /// index of invalid chunk
+        chunk_index: c_ulong,
+        /// number of validators
+        n_validators: c_ulong,
+    },
```
Comment (Harrm) [inline src/erasure_coding.rs:33]: Hm, how does it work when converted to C header?

File: src/erasure_coding.rs [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    }
+}
+
+/// Cleans the data block
+#[allow(unused_attributes)]
+#[no_mangle]
+pub unsafe extern "C" fn clean_data_block(data: *mut DataBlock) {
+    debug_assert!(!data.is_null());
+    debug_assert!(!(*data).array.is_null());
+    Box::from_raw((*data).array);
```
Comment (Harrm) [inline src/erasure_coding.rs:126]: I don't understand why it's written this way.

File: src/erasure_coding.rs [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    debug_assert!(!chunk_list.is_null());
+    debug_assert!(!(*chunk_list).data.is_null());
+    let data = std::mem::transmute::<*mut ChunksList, &mut ChunksList>(chunk_list);
+    {
+        let chunks = slice::from_raw_parts_mut(data.data, data.count as usize);
+        for chunk in chunks.iter_mut() {
+            clean_chunk(chunk);
+        }
+    }
+    Box::from_raw(data.data);
```
Comment (Harrm) [inline src/erasure_coding.rs:150]: Here, too.

File: src/erasure_coding.rs [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+/// Reconstruct data from a set of chunks.
+///
+/// Provide an iterator containing chunk data and the corresponding index.
+/// The indices of the present chunks must be indicated. If too few chunks
+/// are provided, recovery is not possible.
+///
+/// Works only up to 65536 validators, and `n_validators` must be non-zero.
+#[allow(unused_attributes)]
+#[no_mangle]
+pub unsafe extern "C" fn reconstruct(
```
Comment (Harrm) [inline src/erasure_coding.rs]: Remind me, will these functions be prefixed when exported to a C library header? It's just I think that unprefixed C functions may be troublesome due to name collisions.

File: .gitignore (hunk 1 above)
Comment (iceseer) [reply .gitignore]: Из гитигнора надо убрать

File: src/erasure_coding.rs (hunk 3 above)
Comment (iceseer) [reply src/erasure_coding.rs:33]: Да, там генерится тэг-енум и юнион, содержащий структуру

File: src/erasure_coding.rs (hunk 4 above)
Comment (iceseer) [reply src/erasure_coding.rs:126]: Нам надо вернуть память аллоцированную в расте, очищать ее тоже надо в расте.

File: src/erasure_coding.rs (hunk 5 above)
Comment (iceseer) [reply src/erasure_coding.rs:150]: см. выше

File: src/erasure_coding.rs (hunk 6 above)
Comment (iceseer) [reply src/erasure_coding.rs]: Нет, префикса добавлено не будет.

**Review by Harrm — APPROVED** [review]

**Pull request merged**